### PR TITLE
Reader: Fix the url we send to the follow button in Full Post View

### PR DESCRIPTION
**Pull request by blowery**
This still has a big problem: the following state will be correct on load, but when the first page of results come in from the subscription sync, the following state will be dropped.

We need to change how we consume pages of subs in the feed-subscription-store, merging the page with current results, and possibly resorting the results by sub date.

### CODE REVIEW

**Comment by blowery** [comment]
Repro'ing this is "fun".

To see it, you have to have an account with at least 50 subs, preferably around 1000. Find a one of the first things you subscribed to, open the feed steam, then open a post into full post view.

Reload full post view. 

Watch the follow button start off correct, then flop to the unfollowed state, then flip back to followed once the subscription loads in the background.

**Comment by blowery** [comment]
cc @bluefuton as he may have some insight on how to best fix this

**Comment by blowery** [comment]
flipper-the-button

**Comment by bluefuton** [comment]
Rebased and forced.

I have around 270 feeds and I'm able to repro with one of my early subscriptions, http://calypso.localhost:3000/read/post/feed/44902/935964996. Digging into it now.

File: client/lib/reader-feed-subscriptions/index.js [hunk 1]
```diff
@@ -138,37 +135,18 @@ var FeedSubscriptionStore = {
...
-			};
-		} else {
-			// Looks like the first results we've received...
-			subscriptions = {
-				count: newSubscriptions.size,
-				list: newSubscriptions
-			};
-		}
+		forEach( subscriptionsWithState, function( subscription ) {
+			addSubscription( subscription.toJS(), false );
```
Comment (bluefuton) [inline client/lib/reader-feed-subscriptions/index.js]: The `toJS()` isn't ideal, but addSubscription() currently expects a plain JS object.

**Comment by bluefuton** [comment]
The fix here is to run new pages of subscriptions through `addSubscription()` rather than updating the subscriptions list directly. `addSubscription()` looks for subs that already exist in the store, and updates the existing sub if it finds one. e14980d remedies the 'followed-unfollowed-followed' flip you found @blowery.

File: client/lib/reader-feed-subscriptions/index.js (hunk 1 above)
Comment (blowery) [inline client/lib/reader-feed-subscriptions/index.js]: oof. that's a _lot_ of mutations.

**Comment by bluefuton** [comment]
Todo:
- [ ] Revert to using `total_subscriptions` from the subscriptions list response. Without it, the subscription count increments gradually as the subs are loaded in, which is no good for reporting stats (#3517) and feels a bit odd on Manage Followed Sites, too.

File: client/lib/reader-feed-subscriptions/index.js [hunk 2]
```diff
@@ -138,37 +135,20 @@ var FeedSubscriptionStore = {
...
-			subscriptions = {
-				count: newSubscriptions.size,
-				list: newSubscriptions
-			};
-		}
+		subscriptions.list = subscriptions.list.asMutable();
+		forEach( subscriptionsWithState, function( subscription ) {
+			addSubscription( subscription.toJS(), false );
+		} );
+		subscriptions.list = subscriptions.list.asImmutable();
```
Comment (blowery) [inline client/lib/reader-feed-subscriptions/index.js]: I think we need to tease out what addSubscription actually does into a helper.

**Comment by bluefuton** [comment]
Looks good on the whole, but I found one odd quirk. After viewing a full post for an existing sub, it now pops to the top of the sub list (as seen in Manage Following):

https://cloudup.com/ch5LvM8iKxh

**Comment by blowery** [comment]
Ah, nice catch. Looks like we need to sort the list on sub date client side after accepting new subs

**Comment by blowery** [comment]
Rebased and repushed. Working on fixing the bump to top.

**Comment by blowery** [comment]
OK, ready for review again @bluefuton :)

**Comment by bluefuton** [comment]
Cool! :eyes:

**Comment by bluefuton** [comment]
That's done the trick. :ship:

**Pull request merged**